### DISCUSSION TITLE
What is the workflow for debugging the language server?

### DISCUSSION THREAD

**Discussion by NullVoxPopuli** [typed-ember/glint#634] [Q&A]
caveat, this is on a branch for this PR: https://github.com/typed-ember/glint/pull/615, and I'm trying to verify in-repo (in the gjs test-app in test-packages)

For example, I'm getting "server stopped: timed out" issues, and I'd like to debug what's going on, but atm, I feel very clunky when working with language server development with vscode:
- one terminal: run `yarn build --watch`
- another terminal: manually run `yarn extension:package`
- in vscode: 
  - (initially) uninstall Glint, and install the local *.vsix file
  - restart extension host
  - open the outputs channel
  - select the glint channel
  - restart glint from the commands prompt
  - observe the restart command was sent
  - receive error that glint couldn't start

I've tried adding logs in the constructor of the language server class, but they never print. maybe there is a way to enable verbose logging?

What I'd like to do is use the actual debugger and breakpoints and such. 

Is there a better workflow?
To me, it feels like I'm doing something terribly wrong / dark-ages-esque, and there must be a better way!

**Comment by NullVoxPopuli** [comment]
Found these Glint settings in vscode:
![image](https://github.com/typed-ember/glint/assets/199018/70f97ad6-3da0-4791-9d8f-5c5e66cf707f)


This is an improvement -- it's not exactly what I was looking for, but it allows me to explore more things.

**Reply by NullVoxPopuli** [reply]
even with these options tho, I can't get any of my own logging to come out of the server :thinking:

**Reply by NullVoxPopuli** [reply]
gonna see if this helps https://code.visualstudio.com/api/get-started/your-first-extension

**Comment by dfreeman** [comment]
Like we talked about in Discord, there's a launch config in `.vscode/launch.json` that will allow you to launch the extension in debug mode directly in an "extension development" instance of Code. This lets you skip a bunch of the steps above, like uninstalling/packaging/installing your local development copy.

For the language server itself, though, I tend to debug in the context of a test rather than live, since the live version of the language server runs in a separate process (it _is_ possible to attach a debugger to that, but generally I've found to be more trouble than it's worth).

**Reply by NullVoxPopuli** [reply]
oh right, thanks! I'm not used to a `.vscode/launch.json`, so I forgot that you mentioned that last I was looking at glint. I'll give this a go!

> (it is possible to attach a debugger to that, but generally I've found to be more trouble than it's worth

yeah, it's just the main thing I want to debug is the incorrect-syntax handling of content-tag -- because I somehow broke something, yet all the tests are green. it's very puzzling.

**Reply by NullVoxPopuli** [reply]
the .vscode/launch is doing exactly what I want! most excellent!

I'm getting some error / log output:
![image](https://github.com/typed-ember/glint/assets/199018/d9dc4316-292f-42d6-ab15-b8876c71f7fa)

So now I can hopefully try to figure out how to expose these parsing errors! :tada:

**Reply by NullVoxPopuli** [reply]
one thing I noticed as I've been playing around with this is that diagnostics seem to not be coming through, even though they're correctly recreated. :thinking: I've pushed some stuff up to the PR if you want to take a look

**Reply by NullVoxPopuli** [reply]
funny enough, neovim is providing diagnostics:
![image](https://github.com/typed-ember/glint/assets/199018/459d7375-6ab8-4b0c-89bf-f00faf6397dd)



so it's probably just that I don't know how to use vscode :sweat_smile: 

this error in the screenshot, text here:
```
2. Glint encountered an error computing diagnostics for this file. This is likely a bug in Glint; please file an issue, including any code and/or steps to follow to reproduce the error.
   
   Error: Parse Error at /home/nullvoxpopuli/Development/OpenSource/glint/test-packages/ts-template-imports-app/src/index.gts:1:17: 1:29
       at module.exports.__wbg_Error_3c9a6c51df21f7cb (/home/nullvoxpopuli/Development/OpenSource/glint/node_modules/content-tag/content_tag.js:292:17)
       at wasm://wasm/008696fe:wasm-function[371]:0x13efc6
       at wasm://wasm/008696fe:wasm-function[1205]:0x1bb467
       at Preprocessor.parse (/home/nullvoxpopuli/Development/OpenSource/glint/node_modules/content-tag/content_tag.js:267:18)
       at preprocess (/home/nullvoxpopuli/Development/OpenSource/glint/packages/environment-ember-template-imports/-private/environment/preprocess.js:15:23)
       at parseScript (file:///home/nullvoxpopuli/.local/share/nvim/mason/packages/glint/node_modules/@glint/core/lib/transform/template/rewrite-module.js:67:36)
       at calculateCorrelatedSpans (file:///home/nullvoxpopuli/.local/share/nvim/mason/packages/glint/node_modules/@glint/core/lib/transform/template/rewrite-module.js:36:33)
       at rewriteModule (file:///home/nullvoxpopuli/.local/share/nvim/mason/packages/glint/node_modules/@glint/core/lib/transform/template/rewrite-module.js:16:48)
       at TransformManager.getTransformInfo (file:///home/nullvoxpopuli/.local/share/nvim/mason/packages/glint/node_modules/@glint/core/lib/common/transform-manager.js:243:41)
       at TransformManager.readTransformedFile (file:///home/nullvoxpopuli/.local/share/nvim/mason/packages/glint/node_modules/@glint/core/lib/common/transform-manager.js:56:38)
```

is unexpected --- I was hoping it would be as nice as the `glint` cli output:
```
❯ yarn glint
yarn run v1.22.4
$ /home/nullvoxpopuli/Development/OpenSource/glint/node_modules/.bin/glint
src/index.gts:1:30 - error TS0: Expected ',', got 'string literal (./Greeting, './Greeting')'
   ╭─[/home/nullvoxpopuli/Development/OpenSource/glint/test-packages/ts-template-imports-app/src/index.gts:1:1]
 1 │ import Greeting './Greeting';
   ·                 ────────────
   ╰────

1 import Greeting './Greeting';
                               
2 
  

```
though, there is a but of duplication here -- been trying some message formatting from swc

**Reply by dfreeman** [reply]
I don't have the time to try and debug your PR for you, but I can say from looking over the diff that `console.log` calls in the language server are a recipe for a bad time: Code uses stdout for IPC with language servers, so emitting arbitrary text there will usually make it throw up its figurative hands and assume the server is broken.

**Reply by NullVoxPopuli** [reply]
Thats very helpful information! Thank you!

**Reply by lifeart** [reply]
@NullVoxPopuli diagnostics issue may be related to https://github.com/typed-ember/glint/issues/407
